urlff74b96b0d_HugoFeng_HospitalProject_tgz-pJ8-hugo_project_hospital_ProjectTestJ8/src/hugo/project/hospital/Hospital.java

```diff
@@ -0,0 +1,655 @@
+						if (!multiWard.isFull()) {
+							multiWard.addPatient(patient);
+							reception.addPatientInfo(patient, this, multiWard);
+							waitingRoom.waitingList.remove(patient);
+							return;
+						}
+					}
+					// If none available, find a ward for single
+					for (Ward singleWard : wardsForSingle) {
+						if (!singleWard.isFull()) {
+							singleWard.addPatient(patient);
+							reception.addPatientInfo(patient, this, singleWard);
+							waitingRoom.waitingList.remove(patient);
+							return;
+						}
+					}
+				}
+			}
+		}
+
+		
+		public void signOutPatient(PatientInfo info) {
+			Room room = info.getRoom();
+			room.removePatient(info.getPatient());
+			if(!waitingRoom.noOneWaiting()) addPatientFromWaitingRoom();
+		}
+		
+		public void signOutPatient(Patient p) {
+			signOutPatient(reception.getPatientInfo(p));
+		}
+		
+		public Device hasDevice(String type) {
+			Device targetDevice = deviceFactory(type, 0);
+			for (Device device : deviceList) {
+				if(device.equals(targetDevice) && device.getAmount()>0) return device;
+			}
+			return null;
+		}
+		
+		public void addDevice(String type, int addAmount) {
+			Device newDevice = hasDevice(type);
+			if(newDevice != null) newDevice.addAmount(addAmount);
+			else deviceList.addFirst(deviceFactory(type, addAmount));
+		}
+		
+		public void removeDevice(String type) {
+			Device targetDevice = hasDevice(type);
+			if(targetDevice != null) deviceList.remove(targetDevice);
+		}
+		
+		public void removeDevice(String type, int amount) {
+			Device targetDevice = hasDevice(type);
+			if(targetDevice != null) targetDevice.reduceAmount(amount);
+		}
+		
+
+		public int compareTo(Department dep) {
+			return this.name.compareTo(dep.name);
+		}
+		
+		public String toString() {
+			return info();		
+		}
+		
+		public String getName() {
+			return name;
+		}
+
+		public String info() {
+			String roomInfoListString = "";
+			String deviceInfoString = "  Device: [ ";
+			for (Ward room : wardsForSingle) {
+				roomInfoListString += room.toString();
+			}
+			for (Ward room : wardsForMulti) {
+				roomInfoListString += room.toString();
+			}
+			for (Device device : deviceList) {
+				deviceInfoString += device.getType() + "(" + device.getAmount() + ") ";
+			}
+			deviceInfoString += "]";
+			return this.getName() +":\n" 
+			+ deviceInfoString + "\n"
+			+ waitingRoom.toString() + "\n"
+			+ roomInfoListString + "\n";
+		}
+	}
+	
+	protected class NeurologyDepartment extends Department {
+		public NeurologyDepartment() {
+			name = "Neurology";
+			waitingRoom = new WaitingRoom("WR " + name);
+			wardsForMulti = new LinkedList<Ward>();
+			wardsForSingle = new LinkedList<Ward>();
+			wardsForMulti.addLast(new Ward(107, 2));
+			wardsForMulti.addLast(new Ward(108, 2));
+			wardsForMulti.addLast(new Ward(109, 2));
+			wardsForSingle.addLast(new Ward(101, 1));
+			wardsForSingle.addLast(new Ward(102, 1));
+			deviceList = new LinkedList<Device>();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 410:419, 486:495.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.